prereise/gather/griddata/hifld/data_process/topology.py

```diff
@@ -0,0 +1,115 @@
+    cc2_adj = cc1_neighbors & cc2_all_state
+    return cc1_adj, cc2_adj
+
+
+def connect_islands_with_minimum_cost(
+    lines, subs, cost_metric=haversine, state_neighbor=None
+):
+    """Connect a group of islands defined by lines and substations into one big
+    island with minimum cost.
+
+    :param pandas.DataFrame lines: line data frame indexed by ``ID`` with two columns
+        ``SUB_1_ID`` and ``SUB_2_ID`` representing the corresponding substations at the
+        start and end points.
+    :param pandas.DataFrame subs: substation data frame indexed by ``ID`` with two
+        columns ``LATITUDE`` and ``LONGITUDE`` representing the geographical
+        coordination of each entry.
+    :param function cost_metric: a function defines the cost metric to calculate the
+        weight of lines, defaults to :py:func:`haversine`.
+    :param dict state_neighbor: a dictionary defines the adjacency relationship among
+        states, defaults to None and the constant dictionary ``abv_state_neighbor``
+        defined in the ``const`` module is used.
+    :return: (*tuple*) -- a pair of lists, the first one is a list of all potential
+        line candidates among given connected components; the second one is a list
+        of subsequence of the first entry, representing the chosen lines to form a
+        minimum spanning tree. Each entry of both lists is a 3-tuple, index of the
+        first connected component, index of the second connected component,
+        a dictionary with keys containing ``start``, ``end`` and ``weight``, defines
+        the ``from substation ID``, ``to substation ID`` and the weight of the line
+        calculated by the ``cost_metric``.
+
+        .. note:: the indexes of connected components are defined by the size,
+            i.e. number of nodes, of the connected components in ascending order.
+    """
+    if state_neighbor is None:
+        state_neighbor = abv_state_neighbor
+
+    # Construct a networkx graph using lines data frame
+    graph = nx.convert_matrix.from_pandas_edgelist(lines, "SUB_1_ID", "SUB_2_ID")
+
+    # Get the full list of connected components of the resultant graph sored by size in
+    # ascending order
+    all_cc = sorted(list(nx.connected_components(graph)), key=len)
+
+    # Loop through all the combinations of connected components to identify potential
+    # edges filtered by the adjacency relationship specified by the user
+    edge_list = []
+    for ind1, ind2 in tqdm(combinations(range(len(all_cc)), r=2)):
+        cc1 = all_cc[ind1]
+        cc2 = all_cc[ind2]
+        # Find the adjacency states within the two connected components against each
+        # other
+        cc1_adj, cc2_adj = find_adj_state_of_2_conn_comp(cc1, cc2, state_neighbor, subs)
+        if len(cc1_adj) > 0:
+            nodes1 = subs.loc[cc1].query("STATE in @cc1_adj")[["LATITUDE", "LONGITUDE"]]
+            nodes2 = subs.loc[cc2].query("STATE in @cc2_adj")[["LATITUDE", "LONGITUDE"]]
```

Comment:
`query` inside a big loop always makes me nervous. If we outside the loop call:
```python
state_subs = pd.Series(subs.groupby("STATE").groups).map(set)
```
and in the loop call:
```python
nodes1 = subs.loc[
    subs.index.isin(cc1 & set().union(*[state_subs[s] for s in cc1_adj]),
    ["LATITUDE", "LONGITUDE"]
]
```
does that improve performance at all?

Reply:
Yeah, I used chain indexing instead of query, which is kind of equivalent here and again no noticeable performance improvement. `query` is supposed to be slower due to string evaluation, but not that bad, reference post [here](https://stackoverflow.com/questions/49936557/pandas-dataframe-loc-vs-query-performance/49937318).